nntrainer/layers/lstm.cpp

```diff
@@ -74,7 +138,72 @@ void LSTMLayer::setRecurrentActivation(ActivationType activation) {
 }
 
 void LSTMLayer::forwarding(bool training) {
-  // NYI
+  Tensor &weight_xh =
+    weightAt(static_cast<int>(LSTMParams::weight_xh)).getVariableRef();
+  Tensor &weight_hh =
+    weightAt(static_cast<int>(LSTMParams::weight_hh)).getVariableRef();
+  Tensor &bias_h =
+    weightAt(static_cast<int>(LSTMParams::bias_h)).getVariableRef();
+
+  Tensor &hidden_ = net_hidden[0]->getVariableRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+  Tensor &m_cell_ = mem_cell->getVariableRef();
+
+  Tensor temp;
+  Tensor hs_prev;
+  Tensor cs_prev;
+  Tensor hs;
+  Tensor cs;
+  Tensor f;
+  Tensor g;
+  Tensor i;
+  Tensor o;
+
+  for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
+    Tensor islice = input_.getBatchSlice(b, 1);
+    Tensor oslice = hidden_.getBatchSlice(b, 1);
+    Tensor cell = m_cell_.getBatchSlice(b, 1);
+
+    for (unsigned int t = 0; t < islice.height(); ++t) {
+      Tensor xs = input_.getSharedDataTensor(TensorDim(1, 1, 1, islice.width()),
+                                             t * islice.width());
+      hs = oslice.getSharedDataTensor(TensorDim(1, 1, 1, oslice.width()),
+                                      t * oslice.width());
+      cs = cell.getSharedDataTensor(TensorDim(1, 1, 1, cell.width()),
+                                    t * cell.width());
+
+      if (t > 0) {
+        hs_prev = oslice.getSharedDataTensor(TensorDim(1, 1, 1, oslice.width()),
+                                             (t - 1) * oslice.width());
+        cs_prev = cell.getSharedDataTensor(TensorDim(1, 1, 1, cell.width()),
+                                           (t - 1) * cell.width());
+      } else {
+        hs_prev = h_prev.getBatchSlice(b, 1);
+        cs_prev = c_prev.getBatchSlice(b, 1);
+      }
+      hs_prev.dot(weight_hh, temp);
+      temp.add_i(bias_h);
+      temp.add_i(xs.dot(weight_xh));
+
+      f = temp.getSharedDataTensor(TensorDim(1, 1, 1, unit), 0);
+      g = temp.getSharedDataTensor(TensorDim(1, 1, 1, unit), temp.width());
+      i = temp.getSharedDataTensor(TensorDim(1, 1, 1, unit), 2 * temp.width());
+      o = temp.getSharedDataTensor(TensorDim(1, 1, 1, unit), 3 * temp.width());
+
+      recurrent_acti_func.run_fn(f, f);
+      recurrent_acti_func.run_fn(i, i);
+      recurrent_acti_func.run_fn(o, o);
+      acti_func.run_fn(g, g);
```

Comment:
if an activation function is not element-wise, this would break. But I wonder if there is any case for this.

Reply:
softmax is not element-wise, but I doubt someone would use it here though. Add it as a note though or restrict to be only ReLU, TanH, Sigmoid.

Reply:
yes. there will be the limitation about the activation in later PR. only element wise activation is working.